Create draft CONTRIBUTING.md (#151)
GreenScheduler/cats: docs/source/contributing.rst

established by the existing CATS codebase. CATS has a fairly comprehensive test suite that runs
automatically against all pull requests and new code should either come with new tests or with an
explanation about why tests for the new code are not included. Please indicate where changes to 
behavior have been made (especially where this means changes to the tests have also been needed).  
CATS includes documentation which should be updated by the pull requests making changes to the
code (although documentation only pull requests are welcome). Some of this documentation is
automatically generated (from doc strings and help text for command line tools) so please make
sure that this internal documentation is up to date.

Testing can also be undertaken in an isolated environment prior to making a pull request and this
can make code development significantly easer. We run tests using ``flake8`` for basic linting,
``pytest`` for the majority of unit and integration tests, and ``mypy`` to check type annotations
and for the static analysis this permits. In a checked out copy of the source, the following installs
the prerequisites and runs all tests::

>   python3 -m pip install '.[test]'
>   python3 -m pip install flake8
>   python3 -m pip install '.[types]'
>   flake8 . --count --select=E9,F63,F7,F82 --show-source
>   python3 -m mypy cats
>   python3 -m pytest

Reviewer: Why do we need `python3 -m` prepended to the commands? Shouldn't plain `pip install ...` for the first three and then e.g. `mypy cats` work (I think we can assume nobody will be attempting to use Python 2 anymore, if so then god help them)?

```suggestion
  python3 -m pip install '.[test]'
  python3 -m pip install flake8
  python3 -m pip install '.[types]'
  flake8 . --count --select=E9,F63,F7,F82 --show-source
  python3 -m mypy cats
  python3 -m pytest
```

> Updated documentation is automatically generated for PRs and will be available from a link within
> the pull request. 

Reviewer: I misunderstood what this implied on first read, so imagine users could too - to me it is stating that new documentation gets written for you e.g. by AI somehow, whereas now I realise you mean any PR docs updates will be built automatically into a PR branch version of the live docs, right?

So I suggest being more clear/explicit with:

```suggestion
A new build incorporating any updates to the documentation is automatically generated for each PR
and will become available from a link within the pull request. 
```

Making a new release
^^^^^^^^^^^^^^^^^^^^

> Those of us with commit access to the main CATS repository on GitHub are able to generate a new release and publish this to PyPi. This

Reviewer: ```suggestion
Those of us with commit access to the main CATS repository on GitHub are able to generate a new release and publish this to PyPI. This
```

should be discussed ahead of time (via a PR changing the version string, see 1 below) and once broad agreement is in place a release can be created as follows:

 1. Merge a pull request onto main that updates the CATS version number ``__version__`` in ``__init__.py`` and adds any release notes / key changes to the documentation. We use a "major.minor.patch" semantic versioning scheme; for bug fixes etc. bump the patch number, for significant new features bump the minor version number, for changes that break previous behavior update the major version number.
 2. Check that all tests have passed after the merge and that the "latest" documentation at read the docs is updated.

>  3. Create a release via the GitHub web interface. This involves creating a new tag ("v1.2.3" for version "1.2.3"), giving the release a name (just "1.2.3"), and adding short release notes using markdown as needed. make sure this is "set as the latest release".

Reviewer: ```suggestion
 3. Create a release via the GitHub web interface. This involves creating a new tag ("v1.2.3" for version "1.2.3"), giving the release a name (just "1.2.3"), and adding short release notes using markdown as needed. Make sure this is "set as the latest release".
```

>  4. After a short time you should be able to check that the new release exists on PyPi and is documented in the stable docs on read the docs.   

Reviewer: ```suggestion
 4. After a short time you should be able to check that the new release exists on PyPI and is documented in the stable docs on read the docs.   
```
